packages/lib/gpcircuits/test/bounds.spec.ts

```diff
@@ -0,0 +1,268 @@
+  if (
+    bounds.some((bound) => !inInterval(bound, [0n, (1n << BigInt(nBits)) - 1n]))
+  ) {
+    throw new TypeError("Invalid bounds");
+  }
+  const [a, b] = bounds;
+  return (
+    [
+      [1n << BigInt(nBits), (1n << BigInt(nBits)) + a - 1n],
+      [
+        21888242871839275222246405745257275088548364400416034343698204186575808495617n -
+          (1n << BigInt(nBits)) +
+          b +
+          1n,
+        21888242871839275222246405745257275088548364400416034343698204186575808495616n
+      ]
+    ] as [bigint, bigint][]
+  ).filter(([c, d]) => d - c >= 0);
+}
+
+// The complement interval of [0, 1 << n[ ∪ critical interval.
+function complementInterval(
+  bounds: [bigint, bigint],
+  nBits: number
+): [bigint, bigint] {
+  if (bounds.some((bound) => !inInterval(bound, [0n, 1n << BigInt(nBits)]))) {
+    throw new TypeError("Invalid bounds");
+  }
+  const [a, b] = bounds;
+  const c = (1n << BigInt(nBits)) + a;
+  const d =
+    21888242871839275222246405745257275088548364400416034343698204186575808495617n -
+    (1n << BigInt(nBits)) +
+    b;
+  return [c, d];
+}
+
+describe("bounds.BoundsCheckModule should work for intervals with 64-bit unsigned integer bounds", function () {
+  const sampleIntervals: [bigint, bigint][] = [
+    [0n, 0n],
+    [0n, 1n << 4n],
+    [1n << 5n, 1n << 29n],
+    [1n << 4n, 1n << 31n],
+    [0n, (1n << 64n) - 1n]
+  ];
+
+  const sampleValues = [0n, 1n << 5n, 1n << 16n, 1n << 30n, (1n << 64n) - 1n];
+
+  let circuit: WitnessTester<
+    BoundsCheckModuleInputNamesType,
+    BoundsCheckModuleOutputNamesType
+  >;
+
+  this.beforeAll(async () => {
+    circuit = await circomkit.WitnessTester("BoundsCheckModule", {
+      file: "bounds",
+      template: "BoundsCheckModule",
+      params: [64]
+    });
+  });
+
+  it("should yield expected output for 64-bit unsigned values and bounds", async () => {
+    for (const comparisonValue of sampleValues) {
+      for (const [minValue, maxValue] of sampleIntervals) {
+        await circuit.expectPass(
+          { comparisonValue, minValue, maxValue },
+          { out: BigInt(+inInterval(comparisonValue, [minValue, maxValue])) }
+        );
+      }
+    }
+  });
+
+  it("should yield false for large field elements lying in a critical interval", async () => {
+    for (const [minValue, maxValue] of sampleIntervals) {
+      for (const comparisonValue of criticalIntervals(
+        [minValue, maxValue],
+        64
+      ).flat()) {
+        await circuit.expectPass(
+          { comparisonValue, minValue, maxValue },
+          { out: 0n }
+        );
+      }
+    }
+  });
+
+  it("should fail for other field elements", async () => {
+    for (const [minValue, maxValue] of sampleIntervals) {
+      const sampleComplementInterval = complementInterval(
+        [minValue, maxValue],
+        64
+      );
+      for (const comparisonValue of sampleComplementInterval.concat([
+        (sampleComplementInterval[0] + sampleComplementInterval[1]) / 2n
```

Comment:
Similar to the above, targeted testing around the boundaries would be welcome.  I.e. for each end of the complement interval, try X-1, X, X+1.
Testing like this is a good way is to avoid off-by-one errors, which are often hard to catch by code inspection.